core/libp2p/multi/multihash.cpp

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "libp2p/multi/multihash.hpp"
+#include <boost/algorithm/hex.hpp>
+#include <boost/format.hpp>
+
+#include "common/hexutil.hpp"
+#include "libp2p/multi/uvarint.hpp"
+
+using kagome::expected::Error;
+using kagome::expected::Result;
+using kagome::expected::Value;
+
+using kagome::common::Buffer;
+using kagome::common::unhex;
+
+namespace libp2p::multi {
+
+  Multihash::Multihash(HashType type, Hash hash)
+      : hash_{std::move(hash)}, type_{type} {
+    UVarint varint(type);
+    auto bytes = varint.toBytes();
+    data_.put(std::vector<uint8_t>(bytes.begin(), bytes.end()));
+    data_.putUint8(hash_.size());
+    data_.put(hash_.toVector());
+  }
+
+  auto Multihash::create(HashType type, const Hash &hash)
+      -> Result<Multihash, std::string> {
+    if (hash.size() > kMaxHashLength) {
+      constexpr auto s =
+          "The hash is too long: %1% bytes against maximum of %2% bytes";
+      return Error{str(boost::format(s) % hash.size() % kMaxHashLength)};
+    }
+
+    return Value{Multihash{type, hash}};
+  }
+
+  auto Multihash::createFromHex(std::string_view hex)
+      -> Result<Multihash, std::string> {
+    return Buffer::fromHex(hex).match(
+        [](Value<Buffer> &v) { return Multihash::createFromBuffer(v.value); },
+        [&hex](auto e) -> Result<Multihash, std::string> {
```

Comment:
E is copied. Use auto&& and STD::forward

Reply:
I changed it to const Error<std::string>& e and use the string in e in the error message.